### PR TITLE
Group & Ungroup for Structs / KeyTable

### PR DESCRIPTION
**Pull request by jigold**

### CODE REVIEW

File: python/hail/keytable.py [hunk 1]
```diff
@@ -1379,5 +1379,113 @@ def range(cls, n, num_partitions=None):
 
         return KeyTable(Env.hc(), Env.hail().keytable.KeyTable.range(Env.hc()._jhc, n, joption(num_partitions)))
 
+    @handle_py4j
+    @record_method
+    @typecheck_method(columns=oneof(strlike, listof(strlike)))
+    def ungroup(self, columns):
```
Comment (tpoterba) [inline python/hail/keytable.py]: should be varargs, not str or list of str

File: python/hail/keytable.py [hunk 2]
```diff
@@ -1379,5 +1379,113 @@ def range(cls, n, num_partitions=None):
...
+        ...                       types={'B': TStruct(['C0', 'C1'], [TBoolean(), TString()]),
+        ...                              'D': TStruct(['cat', 'dog'], [TInt32(), TInt32()]),
+        ...                              'E': TStruct(['A', 'B'], [TInt32(), TInt32()])})
+
+        The types of each column are ``Int32``, ``Struct``, ``Boolean``, ``Struct`` and ``Struct`` respectively.
+
+        +----+--------------------------+-------+-------------------+--------------+
+        | A  |   B                      |   C   | D                 | E            |
+        +====+==========================+=======+===================+==============+
+        | 32 | {"C0":true,"C1":"hello"} | false | {"cat":5,"dog":7} | {"A":5,"B":7}|
```
Comment (tpoterba) [inline python/hail/keytable.py]: don't use `32`, it's a bit confusing with the `Int32` type

File: python/hail/keytable.py [hunk 3]
```diff
@@ -1379,5 +1379,113 @@ def range(cls, n, num_partitions=None):
...
+        +----+--------------------------+-------+-------------------+--------------+
+
+        Both B and D can be ungrouped. However, E cannot be ungrouped because the names of its
+        fields collide with existing columns in the key table.
```
Comment (tpoterba) [inline python/hail/keytable.py]: can we use name mangling here?

File: python/hail/keytable.py [hunk 4]
```diff
@@ -1379,5 +1379,113 @@ def range(cls, n, num_partitions=None):
...
+    @typecheck_method(columns=oneof(strlike, listof(strlike)))
+    def ungroup(self, columns):
+        """Lifts fields of struct columns as distinct top-level columns.
+
+        **Examples**
+
+        ``kt4`` is a :py:class:`.KeyTable` with five columns: A, B, C, D and E.
+
+        >>> kt4 = hc.import_table('data/kt_example4.tsv', impute=True,
+        ...                       types={'B': TStruct(['C0', 'C1'], [TBoolean(), TString()]),
```
Comment (tpoterba) [inline python/hail/keytable.py]: should be `B0` and `B1`?

File: python/hail/keytable.py [hunk 5]
```diff
@@ -1379,5 +1379,113 @@ def range(cls, n, num_partitions=None):
...
+        | A  |   C0  | C1     |   C   | cat | dog  | E            |
+        +====+=======+========+=======+=====+======+==============+
+        | 32 | true  | "hello"| false | 5   |  7   | {"A":5,"B":7}|
+        +----+-------+--------+-------+-----+------+--------------+
+
+        **Notes**
+
+        Multiple columns can be ungrouped at the same time. However, the names of their fields
+        cannot conflict. The new columns are inserted in the same position in the schema as the
+        original struct column.
```
Comment (tpoterba) [inline python/hail/keytable.py]: Why is this in the same position, while `group` puts them at the end? Column order doesn't usually matter much, so this seems like extra complexity if everything else puts stuff at the end.

File: src/main/scala/is/hail/expr/Type.scala [hunk 6]
```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
   }
 
+  def ungroup(x: Array[String]): (TStruct, (Row) => Row) = {
+    val newFieldNames = mutable.Set(fieldNames.toSet.diff(x.toSet).toArray: _*)
+
+    val structsToUngroup = x.map{ name =>
+      fieldOption(name) match {
+        case None => fatal(s"Struct does not have field with name `$name'.")
+        case Some(fd) =>
+          if (!fd.typ.isInstanceOf[TStruct])
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: replace with match statement

File: src/main/scala/is/hail/expr/Type.scala [hunk 7]
```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+      val result = Array.fill[Any](nNewFields)(null)
+      var localIdx = 0
+      var i = 0
+
+      while (i < nFieldsInit) {
+        if (structsToUngroup.contains(i)) {
+          val (_, q, n) = structsToUngroup(i)
+          val structRow = q(r).asInstanceOf[Row]
+          var j = 0
+          while (j < n) {
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: This doesn't handle null correctly: structRow can be null

File: src/main/scala/is/hail/expr/Type.scala [hunk 8]
```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
     filter(fn)
   }
 
+  def ungroup(x: Array[String]): (TStruct, (Row) => Row) = {
+    val newFieldNames = mutable.Set(fieldNames.toSet.diff(x.toSet).toArray: _*)
+
+    val structsToUngroup = x.map{ name =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: needs formatting

File: src/main/scala/is/hail/expr/Type.scala [hunk 9]
```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+      fieldOption(name) match {
+        case None => fatal(s"Struct does not have field with name `$name'.")
+        case Some(fd) =>
+          if (!fd.typ.isInstanceOf[TStruct])
+            fatal(s"Expecting a type of TStruct for field `$name', but found ${ fd.typ.toPrettyString() } ")
+
+          val s = fd.typ.asInstanceOf[TStruct]
+          val fieldsToAdd = s.fields.map(_.name).toArray
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: you added fieldNames above, use that here.

File: src/main/scala/is/hail/expr/Type.scala [hunk 10]
```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+
+          (fd.index, (s, query(name), s.fields.length))
+      }
+    }.toMap
+
+    val nFieldsInit = fields.length
+    val nFieldsToInsert = structsToUngroup.map { case (_, (_, _, n)) => n }.sum
+    val nNewFields = nFieldsInit + nFieldsToInsert - structsToUngroup.size
+
+    val newFields = Array.fill[Field](nNewFields)(null)
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: I'm finding this approach somewhat hard to parse. There's interplay between i and localIdx, but it's not checked. 
A functional thing that does something like
```scala 
fields.flatMap(f => (the field or its subfields))
    .zipWithIndex
    .map { case (f, i) => f.copy(index = i) }
```
may be easier to understand

File: src/main/scala/is/hail/expr/Type.scala [hunk 11]
```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+      }
+
+      Row.fromSeq(result)
+    }
+
+    (TStruct(newFields), ungrouper)
+  }
+
+  def group(dest: String, names: Array[String]): (TStruct, (Row) => Row) = {
+    val fieldsToGroup = names.zipWithIndex.map{ case (name, i) =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: space between map and {

File: src/main/scala/is/hail/expr/Type.scala [hunk 12]
```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+      fieldOption(name) match {
+        case None => fatal(s"Struct does not have field with name `$name'.")
+        case Some(fd) => (fd.index, (i, fd))
+      }
+    }.toMap
+
+    val keepIndices = fields.filter(fd => !fieldsToGroup.contains(fd.index) && fd.name != dest).map(_.index).toArray
+
+    val nNewFields = keepIndices.length + 1
+    val newFields = Array.fill[Field](nNewFields)(null)
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: This is also hard to parse. How about something like:
```scala
val keptFields = (fields.filter(fd => ...) ++ Array(the grouped thing))
.zipWithIndex
.map { case (f, i) => f.copy(index = i) }
```

File: src/main/scala/is/hail/expr/Type.scala [hunk 13]
```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+
+    var localIdx = 0
+    keepIndices.foreach { i =>
+      newFields(localIdx) = fields(i).copy(index = localIdx)
+      localIdx += 1
+    }
+
+    newFields(localIdx) = Field(dest, TStruct(fieldsToGroup.map { case (_, (i, fd)) => Field(fd.name, fd.typ, i) }.toIndexedSeq), localIdx)
+
+    val grouper = (r: Row) => {
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: image

File: src/main/scala/is/hail/expr/Type.scala [hunk 14]
```diff
@@ -1469,6 +1472,115 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+      val result = Array.fill[Any](nNewFields)(null)
+      var localIdx = 0
+
+      keepIndices.foreach { i =>
+        result(localIdx) = r.get(i)
+        localIdx += 1
+      }
+
+      result(localIdx) = Row(fieldsToGroup.map { case (idx, _) => r.get(idx)}.toSeq: _*)
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: assert that localIdx == nNewFields - 1

File: src/test/scala/is/hail/methods/KeyTableSuite.scala [hunk 15]
```diff
@@ -413,4 +413,100 @@ class KeyTableSuite extends SparkSuite {
...
+    val signatureExpected = TStruct(("Sample", TString), ("1", TFloat32), ("2", TString),
+      ("field1", TInt32), ("field2", TInt32),
+      ("a", TInt32), ("b", TString), ("c", TFloat64),
+      ("field4", TStruct(("field0", TInt32), ("field1", TString), ("field2", TFloat64))),
+      ("field5", TStruct(("a", TInt32), ("b", TString), ("c", TFloat64))))
+    val rddExpected = sc.parallelize(dataExpected.map(Row.fromSeq(_)))
+    val ktExpected = KeyTable(hc, rddExpected, signatureExpected)
+    assert(kt.ungroup(Array("field0", "field3")).same(ktExpected))
+
+    intercept[HailException](kt.ungroup(Array("field4")))
```
Comment (tpoterba) [inline src/test/scala/is/hail/methods/KeyTableSuite.scala]: use TestUtils.interceptFatal, and part of the message you expect. It's hard to see on just reading this which kind of failure is expected

File: src/test/scala/is/hail/methods/KeyTableSuite.scala [hunk 16]
```diff
@@ -413,4 +413,100 @@ class KeyTableSuite extends SparkSuite {
...
+
+    val rdd3 = sc.parallelize(data3.map(Row.fromSeq(_)))
+    val rddExpected3 = sc.parallelize(expectedData3.map(Row.fromSeq(_)))
+
+    val kt3 = KeyTable(hc, rdd3, TStruct(("A", TInt32), ("B", TStruct(("a", TInt32), ("b", TStruct(("i", TString))))), ("C", TBoolean)))
+    val ktExpected3 = KeyTable(hc, rddExpected3, TStruct(("A", TInt32), ("B", TStruct(("a", TInt32), ("i", TString))), ("C", TBoolean)))
+
+    assert(kt3.annotate("B = ungroup(B, b)").same(ktExpected3))
+
+    intercept[HailException](kt3.annotate("B = ungroup(B)"))
```
Comment (tpoterba) [inline src/test/scala/is/hail/methods/KeyTableSuite.scala]: same as above: TestUtils.interceptFatal

File: src/test/scala/is/hail/methods/KeyTableSuite.scala [hunk 17]
```diff
@@ -413,4 +413,100 @@ class KeyTableSuite extends SparkSuite {
...
+    val rdd2 = sc.parallelize(data2.map(Row.fromSeq(_)))
+    val kt2 = KeyTable(hc, rdd2, TStruct(("foo", TStruct(("a", TString), ("b", TInt32)))))
+
+    val dataExpected = Array(Array(Row("Sample1", 5), Row(Row(5, "Sample1"))))
+    val sigExpected = TStruct(("foo", TStruct(("a", TString), ("b", TInt32))), ("X", TStruct(("a", TStruct(("b", TInt32), ("a", TString))))))
+    val kt2Expected = KeyTable(hc, sc.parallelize(dataExpected.map(Row.fromSeq(_))), sigExpected)
+
+    assert(kt2.annotate("X = group(foo, a, b, a)").same(kt2Expected))
+
+    intercept[HailException](kt.annotate("X = group(foo, a, b, b)"))
```
Comment (tpoterba) [inline src/test/scala/is/hail/methods/KeyTableSuite.scala]: same

File: python/hail/keytable.py [hunk 18]
```diff
@@ -1379,5 +1379,113 @@ def range(cls, n, num_partitions=None):
...
+        :rtype: :class:`.KeyTable`
+        """
+
+        return KeyTable(self.hc, self._jkt.ungroup(wrap_to_list(columns)))
+
+    @handle_py4j
+    @record_method
+    @typecheck_method(dest=strlike,
+                      columns=oneof(strlike, listof(strlike)))
+    def group(self, dest, columns):
```
Comment (tpoterba) [inline python/hail/keytable.py]: I thought we were planning to push people away from nested structures as much as possible. I imagine it would be possible to do this in 0.2 with a syntax like:
```python
kt = kt.with(x=Struct(a=kt.a, b=kt.b)).drop(kt.a, kt.b)
```
I definitely see the value of ungroup, but group seems like it is going to collide with expr.

Comment (tpoterba) [reply python/hail/keytable.py]: shall we talk about this in person?

File: python/hail/keytable.py (hunk 1 above)
Comment (jigold) [reply python/hail/keytable.py]: How do I typecheck varargs? Also, this should go on the to-do list for 0.2 as there are other methods that have this interface.

Comment (tpoterba) [reply python/hail/keytable.py]: fair enough.

You can typecheck varargs as a `tupleof(whatever)`

File: python/hail/keytable.py (hunk 3 above)
Comment (jigold) [reply python/hail/keytable.py]: I think it might be confusing to do the mangling as in the duplicate samples case. What do you think of the mangling deduplication function to be the full path (same as flatten). For example, I am worried about this case:

```
Struct{
    pValue: Float64,
    OR: Float64,
    linreg: Struct{
        pValue: Float64,
        OR: Float64
    },
    logreg: Struct{
        pValue: Float64,
        OR: Float64
    }
}
```

becoming

```
Struct{
    pValue: Float64,
    OR: Float64,
    pValue_1: Float64,
    OR_1: Float64
    pValue_2: Float64,
    OR_2: Float64
}
```

I think this is better:
```
Struct{
    pValue: Float64,
    OR: Float64,
    `linreg.pValue`: Float64,
    `linreg.OR`: Float64
    `logreg.pValue`: Float64,
    `logreg.OR`: Float64
}
```

**Comment by jigold** [comment]
@tpoterba back to you.

File: python/hail/keytable.py [hunk 19]
```diff
@@ -1421,4 +1421,114 @@ def maximal_independent_set(self, i, j):
...
+        | A  |   B                      |   C   | D                 | E.A   |  E.B |
+        +====+==========================+=======+===================+==============+
+        | 15 | {"B0":true,"B1":"hello"} | false | {"cat":5,"dog":7} | 5     |  7   |
+        +----+--------------------------+-------+-------------------+--------------+
+
+        **Notes**
+
+        The ungrouped columns are always appended to the end of the table.
+
+        :param column: Names of struct column to ungroup.
```
Comment (tpoterba) [inline python/hail/keytable.py]: :param str column

File: src/main/scala/is/hail/expr/Type.scala [hunk 20]
```diff
@@ -1469,6 +1473,107 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+          s.fields.flatMap { fd =>
+            (fieldNames.contains(fd.name), mangle) match {
+              case (_, true) => Some((identifier + "." + fd.name, fd.typ))
+              case (false, false) => Some((fd.name, fd.typ))
+              case (true, false) =>
+                overlappingNames += fd.name
+                None
+            }
+          }.toArray
+        case other => fatal(s"Expecting a type of TStruct for field `$identifier', but found ${ other.toPrettyString() } ")
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: needs to be `compact=True`. I think the message is also a bit unclear, and the stuff in the Python docs uses `Struct` instead of `TStruct`. How about:

"Can only ungroup fields of type Struct, but found type ${...}"

File: src/main/scala/is/hail/expr/Type.scala [hunk 21]
```diff
@@ -1469,6 +1473,107 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
 
+  def ungroup(identifier: String, mangle: Boolean = false): (TStruct, (Row) => Row) = {
+    val overlappingNames = mutable.Set[String]()
+
+    val ungroupedFields = fieldOption(identifier) match {
+      case None => fatal(s"Struct does not have field with name `$identifier'.")
+      case Some(x) => x.typ match {
+        case s: TStruct =>
+          s.fields.flatMap { fd =>
+            (fieldNames.contains(fd.name), mangle) match {
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: this is a linear scan of the columns each time. Convert to set above maybe?

File: src/main/scala/is/hail/expr/Type.scala [hunk 22]
```diff
@@ -1469,6 +1473,107 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+                None
+            }
+          }.toArray
+        case other => fatal(s"Expecting a type of TStruct for field `$identifier', but found ${ other.toPrettyString() } ")
+      }
+    }
+
+    if (overlappingNames.nonEmpty)
+      fatal(s"Found ${ overlappingNames.size } ${ plural(overlappingNames.size, "ungrouped field name") } overlapping existing struct field names.\n" +
+        "Use the `mangle=True` option to deduplicate field names:\n  " +
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: a bit unclear. Maybe "Either rename manually or use the 'mangle' option to handle duplicates.\n  Overlapping fields:\n  "

File: src/main/scala/is/hail/expr/Type.scala [hunk 23]
```diff
@@ -1469,6 +1473,107 @@ final case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+        var i = 0
+        while (i < origSize) {
+          if (i != fdIndexToUngroup) {
+            result(localIdx) = r.get(i)
+            localIdx += 1
+          }
+          i += 1
+        }
+
+        val ugr = r.get(fdIndexToUngroup).asInstanceOf[Row]
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: use `getAs`

File: src/main/scala/is/hail/expr/AST.scala [hunk 24]
```diff
@@ -517,6 +527,89 @@ case class Apply(posn: Position, fn: String, args: Array[AST]) extends AST(posn,
 
         `type` = tNew
 
+      case "ungroup" =>
+        if (args.length != 3)
+          parseError(
+            s"""invalid arguments for method `$fn'
+               |  Usage: $fn(Struct, identifier, mangle)
```
[the remark this answers is not in the document]
Comment (jigold) [reply src/main/scala/is/hail/expr/AST.scala]: I started to do that, but it made the code substantially more complicated. I figured since this is for 0.2, the call to the expression language will be from a Python function so we can handle the default argument in Python.

File: src/main/scala/is/hail/expr/Type.scala [hunk 25]
```diff
@@ -68,7 +69,8 @@ object Type {
   def parseMap(s: String): Map[String, Type] = Parser.parseAnnotationTypes(s)
 }
 
-sealed abstract class Type extends Serializable { self =>
+sealed abstract class Type extends Serializable {
```
[the remark this answers is not in the document]
Comment (jigold) [reply src/main/scala/is/hail/expr/Type.scala]: This kept happening when I used "reformat code" in IntelliJ ...

[the remark this answers is not in the document]
Comment (tpoterba) [reply src/main/scala/is/hail/expr/Type.scala]: Mine autoformats it this way too, even if it doesn't seem like the right style. I think we can leave it here, but in the future just autoformat while your code is selected so we don't all have to rebase on each other's formatting changes
